src/main/java/io/asyncer/r2dbc/mysql/QueryFlow.java

```diff
@@ -1227,16 +1234,63 @@ private static String buildStartTransaction(TransactionDefinition definition) {
     }
 }
 
+final class CreateSavepointState extends AbstractTransactionState {
+
+    private static final int START_TRANSACTION = 1;
+
+    private static final int CREATE_SAVEPOINT = 2;
+
+    private CreateSavepointState(ConnectionState state, int tasks, List<String> statements) {
+        super(state, tasks, statements);
+    }
+
+    @Override
+    boolean cancelTasks() {
+        return false;
```

Comment:
If it looks like following:

```java
@Override
boolean cancelTasks() {
    this.tasks = state.isInTransaction() ? CREATE_SAVEPOINT : START_TRANSACTION | CREATE_SAVEPOINT;

    return false;
}
```

Then `TransactionBatchExchangeable` or `TransactionMultiExchangeable` will check transaction state at `subscribe`, and `this.tasks` will be changed to the correct.

This avoids the use of `@Nullable state`, additional `Supplier`, and `assert`.

Or is there something else to consider?

Reply:
Nice feedback.